File: 015.NN_UIViewAnimationPractice/NN_UIViewAnimationPractice/UIViewAnimationPractice/Controllers/NNSpringAnimationController.m
//
//  NNSpringAnimationController.m
//  NN_UIViewAnimationPractice
//
//  Created by 刘朋坤 on 2018/8/11.
//  Copyright © 2018年 刘朋坤. All rights reserved.
//

#import "NNSpringAnimationController.h"

@interface NNSpringAnimationController ()

@end

@implementation NNSpringAnimationController

- (NSArray *)buttonTitleArray {
    return @[@"spring", @"组合"];
}

- (void)buttonClick:(UIButton *)sender {
    switch (sender.tag) {
        case 0:
            [self springAnimation1];
            break;
        case 1:
            [self springAnimation2];
            break;
    }
}

- (void)springAnimation1 {
    CGRect originalRect = self.nnView.headerImage.frame;
    CGRect changeRect = CGRectMake(self.nnView.headerImage.center.x-50, self.nnView.headerImage.center.y - 200, 100, 100);
    [UIView animateWithDuration:1 delay:0 usingSpringWithDamping:0.3 initialSpringVelocity:4 options:UIViewAnimationOptionCurveLinear animations:^{
        self.nnView.headerImage.frame = changeRect;
    } completion:^(BOOL finished) {
        self.nnView.headerImage.frame = originalRect;
    }];
}

- (void)springAnimation2 {
    CGRect originalRect = self.nnView.headerImage.frame;
    CGRect changeRect = CGRectMake(self.nnView.headerImage.center.x-50, self.nnView.headerImage.center.y - 200, 100, 100);
    [UIView animateWithDuration:1 delay:0 usingSpringWithDamping:0.3 initialSpringVelocity:4 options:UIViewAnimationOptionCurveLinear animations:^{
        self.nnView.headerImage.frame = changeRect;
    } completion:^(BOOL finished) {
        [UIView animateWithDuration:1 delay:0.3 usingSpringWithDamping:0.3 initialSpringVelocity:4 options:UIViewAnimationOptionCurveLinear animations:^{
            self.nnView.headerImage.frame = originalRect;
        } completion:nil];
    }];
}

@end

File: 017.NN_UIViewAnimationPractice_swift/NN_UIViewAnimationPractice_swift/UIViewAnimationPractice/Controllers/NNSpringAnimationController.swift
//
//  NNSpringAnimationController.swift
//  NN_UIViewAnimationPractice
//
//  Created by 刘朋坤 on 2019/1/5.
//  Copyright © 2019 刘朋坤. All rights reserved.
//

import UIKit

class NNSpringAnimationController: NNBaseViewController {

    // 按钮点击事件
    override func buttonClick(index: NSInteger) {
        switch index {
        case 0:
            springAnimation1()
            break
        case 1:
            springAnimation2()
            break
        default:
            break
        }
    }
    
    func springAnimation1() {
        let originalRect = self.nnView.headerImage.frame
        let changeRect = CGRect(x: self.nnView.headerImage.center.x-50, y: self.nnView.headerImage.center.y - 200, width: 100, height: 100)
        UIView.animate(withDuration: 1, delay: 0, usingSpringWithDamping: 0.3, initialSpringVelocity: 4, options: UIView.AnimationOptions.curveLinear, animations: {
            self.nnView.headerImage.frame = changeRect
        }) { (finished) in
            self.nnView.headerImage.frame = originalRect
        }
    }
    
    func springAnimation2() {
        let originalRect = self.nnView.headerImage.frame
        let changeRect = CGRect(x: self.nnView.headerImage.center.x-50, y: self.nnView.headerImage.center.y - 200, width: 100, height: 100)
        UIView.animate(withDuration: 1, delay: 0, usingSpringWithDamping: 0.3, initialSpringVelocity: 4, options: UIView.AnimationOptions.curveLinear, animations: {
            self.nnView.headerImage.frame = changeRect
        }) { (finished) in
            UIView .animate(withDuration: 1, delay: 0.3, usingSpringWithDamping: 0.3, initialSpringVelocity: 4, options: UIView.AnimationOptions.curveLinear, animations: {
                self.nnView.headerImage.frame = originalRect
            }, completion: nil)
        }
    }

    // 数组元素
    override func buttonTitleArrayMethod() -> NSArray {
        return ["spring", "组合"]
    }

}
